### PR TITLE
Erc20 in da place (rinkeby first)

### PR DESCRIPTION
**Pull request by vrolland**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -15,8 +15,8 @@ before_script:
 - sleep 3
 - lerna run --scope @requestnetwork/request-network.js testdeploy
 script:
-- lerna run --scope @requestnetwork/request-network.js test
-- lerna run --scope request-network-smart-contracts test
+- travis_wait lerna run --scope @requestnetwork/request-network.js test --stream
```
Comment (romaric-juniet) [inline .travis.yml]: Do you mind removing the travis_wait? It didn't solve the issue.
We can keep the additions of --stream, though

File: packages/requestNetwork.js/src/servicesExternal/erc20-service.ts [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+     * ERC20 contract's abi
+     */
+    public static abiERC20: any = ERC20.abi;
+
+    private static web3Single: Web3Single;
+
+    /**
+     * ERC20 contract's address
+     */
+    private addressERC20: any = ERC20.abi;
```
Comment (romaric-juniet) [inline packages/requestNetwork.js/src/servicesExternal/erc20-service.ts]: Same value as abiERC20?
Also, the init value is overwritten in the contrsuctor

File: packages/requestNetwork.js/src/servicesExternal/erc20-service.ts [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    /**
+     * Erc20 contract's web3 instance
+     */
+    private instanceERC20: any;
+
+    /**
+     * Constructor to Instantiates a new Erc20Service
+     * @param   _address       address of the ERC20 token
+     */
+    public constructor(_address: string) {
```
Comment (romaric-juniet) [inline packages/requestNetwork.js/src/servicesExternal/erc20-service.ts]: Assert the address has been given in input?

File: packages/requestNetwork.js/src/servicesExternal/erc20-service.ts [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        Erc20Service.web3Single.broadcastMethod(
+            method,
+            (hash: string) => {
+                return promiEvent.eventEmitter.emit('broadcasted', {transaction: {hash}});
+            },
+            (receipt: any) => {
+                // we do nothing here!
+            },
+            async (confirmationNumber: number, receipt: any) => {
+                if (confirmationNumber === _options.numberOfConfirmation) {
```
Comment (romaric-juniet) [inline packages/requestNetwork.js/src/servicesExternal/erc20-service.ts]: `>=` ?

File: packages/requestNetwork.js/test/deploy.ts [hunk 5]
```diff
@@ -77,6 +87,78 @@ web3Single.getDefaultAccount().then((creator) => {
                 addressRequestEthereum = newContractInstance.options.address;
                 newContractInstanceRequestEthereum = newContractInstance;
 
+
+                instanceERC20TestToken.deploy({
+                    data: TestToken.bytecode,
+                    arguments: ['100000000000000000000000000000000000000000000000000000000']
```
Comment (romaric-juniet) [inline packages/requestNetwork.js/test/deploy.ts]: Avoir hardcoded value. We can declare it in a `const`

File: packages/requestNetwork.js/test/deploy.ts [hunk 6]
```diff
@@ -77,6 +87,78 @@ web3Single.getDefaultAccount().then((creator) => {
...
+                })
+                .send({
+                    from: creator,
+                    gas: 1500000
```
Comment (romaric-juniet) [inline packages/requestNetwork.js/test/deploy.ts]: Same, let's put it in a constant on top of the file to easily change it

File: packages/requestNetwork.js/test/deploy.ts [hunk 7]
```diff
@@ -77,6 +87,78 @@ web3Single.getDefaultAccount().then((creator) => {
...
+                }, (error: Error, transactionHash: string) => {
+                    if (error) {
+                        console.log('TestToken - error transactionHash ##########################')
+                        console.log(error)
+                        console.log(transactionHash)
+                        console.log('TestToken - error transactionHash ##########################')
+                    }
+                    // console.log('RequestCore - transactionHash : '+transactionHash);
```
Comment (romaric-juniet) [inline packages/requestNetwork.js/test/deploy.ts]: Remove commented out code

File: packages/requestNetwork.js/test/deploy.ts [hunk 8]
```diff
@@ -77,6 +87,78 @@ web3Single.getDefaultAccount().then((creator) => {
...
+                })
+                .on('error', (error: Error) => {
+                    console.log('TestToken - error transactionHash ##########################')
```
Comment (romaric-juniet) [inline packages/requestNetwork.js/test/deploy.ts]: I don't understand the error message. Why is there "error transactionHash"?

File: packages/requestNetwork.js/test/deploy.ts [hunk 9]
```diff
@@ -77,6 +87,78 @@ web3Single.getDefaultAccount().then((creator) => {
...
+                    console.log(error)
+                    console.log('TestToken - error transactionHash ##########################')
+                })
+                .then( (newContractInstance: any) => {
+                    console.log('TestToken - address : ' + newContractInstance.options.address);
```
Comment (romaric-juniet) [inline packages/requestNetwork.js/test/deploy.ts]: swap lines and console.log the variable

File: packages/requestNetwork.js/test/deploy.ts [hunk 10]
```diff
@@ -77,6 +87,78 @@ web3Single.getDefaultAccount().then((creator) => {
...
+                    addressCentralBank = newContractInstance.options.address;
+
+                    instanceRequestERC20.deploy({
+                        data: requestERC20Json.bytecode,
+                        arguments: [addressRequestCore, addressContractBurner, addressCentralBank]
```
Comment (romaric-juniet) [inline packages/requestNetwork.js/test/deploy.ts]: Explain why we use CentralBank

File: packages/requestNetworkSmartContracts/README.md [hunk 11]
```diff
@@ -27,6 +28,7 @@ truffle deploy --network development
 Contract addresses
 * RequestCore: 0x8fc2e7f2498f1d06461ee2d547002611b801202b
 * RequestEthereum: 0xd88ab9b1691340e04a5bbf78529c11d592d35f57
+* RequestERC20 token CTBK : 0xc3ba385addea98bb0af084b1e8bdc909f0215bbf  (test token CTBK 0x995d6a8c21f24be1dd04e105dd0d83758343e258)
```
Comment (romaric-juniet) [inline packages/requestNetworkSmartContracts/README.md]: Explain more what the relationship between the 2 contract is

File: packages/requestNetworkSmartContracts/migrations/2_deploy_contracts.js [hunk 12]
```diff
@@ -27,11 +28,11 @@ function createInstances() {
   });
 }
 
-function setupContracts({ ethereum: ethereumInstance, core: coreInstance }) {
+function setupContracts({ ethereum: ethereumInstance, core: coreInstance, erc20: erc20Instance }) {
   return Promise.all([
     coreInstance.adminAddTrustedCurrencyContract(ethereumInstance.address),
     ethereumInstance.setFeesPerTenThousand(feesPerTenThousand),
-    ethereumInstance.setMaxCollectable(maxFees)
+    ethereumInstance.setMaxCollectable(maxFees),
```
Comment (romaric-juniet) [inline packages/requestNetworkSmartContracts/migrations/2_deploy_contracts.js]: There's an issue here:
The param `erc20Instance` is given line 18. It comes from the return of `createInstances`. `createInstances` doesn't return the erc20 contract

File: packages/requestNetwork.js/src/servicesExternal/erc20-service.ts (hunk 2 above)
Comment (vrolland) [reply packages/requestNetwork.js/src/servicesExternal/erc20-service.ts]: good point : `private addressERC20: string;`

File: packages/requestNetwork.js/src/servicesExternal/erc20-service.ts (hunk 4 above)
Comment (vrolland) [reply packages/requestNetwork.js/src/servicesExternal/erc20-service.ts]: === to launch is code only once

File: packages/requestNetwork.js/test/deploy.ts (hunk 8 above)
Comment (vrolland) [reply packages/requestNetwork.js/test/deploy.ts]: was a copy/past, I will change it

**Pull request merged**